ospfd: preparation for TI-LFA

GalaxyGorilla opened the pull request:
This PR is is split into 3 parts/commits:

1. A rather cosmetic cleanup of SPF code, no functional changes (might be a bit tricky to review)
2. Introduction of a 'dry run' mode to the SPF code, more about this below
3. Some print helper utility to print route tables and SPF trees

Background: I'm working on TI-LFA (which, in a nutshell, is about generating backup paths in case of failure scenarios) for OSPF and a major part of the involved work concerns so called P and Q spaces which are used to generate backup paths. To generate these spaces one needs to execute the SPF algorithm for arbitrary roots, e.g. not just the local (calculating) router itself. For this purpose this PR introduces a 'dry run' mode which can be used to execute SPF for arbitrary nodes without touching unwanted or unneeded things like local interface data. Later on this mode is supposed to be extended to OSPF's post-processing of the SPF (e.g. inter-area dependencies).

I also have a little test framework for SPF here in place which is supposed to be part of the unit tests later on (first unit tests for OSPF, yay).

brotherdust commented:
@GalaxyGorilla This is an interesting feature! Since I'm the curious type, I tried looking up the RFC's for this and there seem to be quite a few. In reference to this feature, is there particular RFC number(s) you're working from as part of your implementation?

GalaxyGorilla commented:
> @GalaxyGorilla This is an interesting feature! Since I'm the curious type, I tried looking up the RFC's for this and there seem to be quite a few. In reference to this feature, is there particular RFC number(s) you're working from as part of your implementation?

@brotherdust Yes, there's just a draft RFC (or multiple versions of it) out there. Personally I'm using the books 'Segment Routing I & II' for this, the authors of these books are also participating in the RFC.

GalaxyGorilla commented:
@odd22 It would be very nice if you could review the 4th commit I added recently. This commit fixes issues that popped up with the topotest `ospf-topo2`. Maybe there is a more elegant way to solve this, I really don't like the current solution :).

The main problem in this topotest is that two OSPF neighbours are not connected on the same subnet and hence the reverse lookup I introduced in the second commit doesn't work anymore.

odd22 commented:
@GalaxyGorilla Your solution is fine for me even if it is not optimal nor what you are expected. But, for unnumbered interface, there not many solution and I'm afraid that there no other solution which not look to ospf interface.
